cpp/include/raft/core/temporary_device_buffer.hpp

```diff
@@ -0,0 +1,220 @@
+
+  constexpr temporary_device_buffer(temporary_device_buffer&&) = default;
+  constexpr temporary_device_buffer& operator=(temporary_device_buffer&&) = default;
+
+  /**
+   * @brief Construct a new temporary device buffer object
+   *
+   * @param handle raft::device_resources
+   * @param data input pointer
+   * @param extents dimensions of input array
+   * @param write_back if true, any writes to the `view()` of this object will be copid
+   *                   back if the original pointer was in host memory
+   */
+  temporary_device_buffer(device_resources const& handle,
+                          ElementType* data,
+                          Extents extents,
+                          bool write_back = false)
+    : stream_(handle.get_stream()),
+      original_data_(data),
+      extents_{extents},
+      write_back_(write_back),
+      length_([this]() {
+        std::size_t length = 1;
+        for (std::size_t i = 0; i < extents_.rank(); ++i) {
+          length *= extents_.extent(i);
+        }
+        return length;
+      }()),
+      device_id_{get_device_for_address(data)}
+  {
+    if (device_id_ == -1) {
+      typename owning_device_buffer::mapping_type layout{extents_};
+      typename owning_device_buffer::container_policy_type policy{handle.get_stream()};
+
+      owning_device_buffer device_data{layout, policy};
+      raft::copy(device_data.data_handle(), data, length_, handle.get_stream());
+      data_ = data_store{std::in_place_index<1>, std::move(device_data)};
+    } else {
+      data_ = data_store{std::in_place_index<0>, data};
+    }
+  }
+
+  ~temporary_device_buffer() noexcept(is_const_pointer_)
+  {
+    // only need to write data back for non const pointers
+    // when write_back=true and original pointer is in
+    // host memory
+    if constexpr (not is_const_pointer_) {
+      if (write_back_ && device_id_ == -1) {
+        raft::copy(original_data_, std::get<1>(data_).data_handle(), length_, stream_);
+      }
+    }
+  }
+
+  /**
+   * @brief Returns a `raft::device_mdspan`
+   *
+   * @return raft::device_mdspan
+   */
+  auto view() -> view_type
+  {
+    if (device_id_ == -1) {
+      return std::get<1>(data_).view();
+    } else {
+      return make_mdspan<ElementType, index_type, LayoutPolicy, false, true>(original_data_,
+                                                                             extents_);
+    }
+  }
+
+ private:
+  rmm::cuda_stream_view stream_;
+  ElementType* original_data_;
+  data_store data_;
+  Extents extents_;
+  bool write_back_;
+  std::size_t length_;
+  int device_id_;
+};
+
+/**
+ * @brief Factory to create a `raft::temporary_device_buffer`
+ *
+ * @tparam ElementType type of the input
+ * @tparam IndexType index type of `raft::extents`
+ * @tparam LayoutPolicy layout of the input
+ * @tparam ContainerPolicy container to be used to own device memory if needed
+ * @tparam Extents variadic dimensions for `raft::extents`
+ * @param handle raft::device_resources
```

Comment:
Can we add a brief usage example here? We'll also need to manually add these (ideally create a doxygengroup and link it in to a new `rst` file for `temporary_device_buffer).

Reply:
Should the usage example be for any particular method? Maybe something from `neighbors`?

Reply:
Really, it could just be a copy/paste example of including the `device_resources.hpp`, defining a pointer on host, and then using the pointer to invoke the factory function. (Only the factory functions here  need usage examples).

Reply:
Done!